gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopyableFile.java

```diff
@@ -375,6 +379,35 @@ public static List<OwnerAndPermission> resolveReplicatedOwnerAndPermissionsRecur
     return ownerAndPermissions;
   }
 
+  /**
+   * Compute the correct {@link OwnerAndPermission} obtained from replicating source owner and permissions and applying
+   * the {@link PreserveAttributes} rules for fromPath and every ancestor up to but excluding toPath.
+   * Use permissionMap as a cache to reduce the call to hdfs
+   *
+   * @return A list of the computed {@link OwnerAndPermission}s starting from fromPath, up to but excluding toPath.
+   * @throws IOException if toPath is not an ancestor of fromPath.
+   */
+  public static List<OwnerAndPermission> resolveReplicatedOwnerAndPermissionsRecursivelyWithCache(FileSystem sourceFs, Path fromPath,
+      Path toPath, CopyConfiguration copyConfiguration, Cache<String, OwnerAndPermission> permissionMap)
```

Comment:
why would a caller use this one rather than `resolveReplicatedAncestorOwnerAndPermissionsRecursively` - do we need both?  they're not exactly identical, but there seems material overlap.  and, why return a `Map` there but a `List` here?

Reply:
I leave the previous method for backward compatibility as there are some other places that call this method, and if it does not do the calculation for a bunch of files sharing the same flows, introducing a cache is not necessary there.

Reply:
I won't strongly suggest refactoring these into a single impl, but I do highly recommend you either add a TODO about doing that later or else provide a javadoc comment describing how the two relate and who should use each.

BTW, from your reply above about backward compatibility, I almost wonder whether your stance is that the existing method should get `@Deprecated`